ethcore/sync/src/chain/mod.rs

```diff
@@ -1099,6 +1126,36 @@ impl ChainSync {
 						}
 					}
 
+					// get some peers to give us transaction pool
+					if !self.unfetched_pooled_transactions.is_empty() {
```

Comment:
IMHO the whole code could be rewritten more clearly as follows. I did change the code significantly (even the logic) cause the previous one did not make any sense to me. It could easily get stuck in an endless loop forever.
```rust
// Consider asking the peer for transactions we need.
if self.unfetched_pooled_transactions.is_empty() {
  return;
}
let asking = self.peers.get_mut(&peer_id).as_mut().map(|peer| &mut peer.asking_pooled_transactions);
let mut asking = match asking {
    Some(ref mut asking) => asking,
    // no peer or peer does not support eth65
    None => return;
};
const MAX_ASKING: usize = 256;
// we already have too many requests scheduled, don't do anything.
if asking.len() > MAX_ASKING {
  return;
}
let now = Instant::now();
// we assume that only few transactions will actually be added to `asking`, so we only collect
// hashes to remove.
let mut remove_from_unfetched = vec![];
for (ref hash, ref mt item) in self.unfetched_pooled_transactions.iter_mut() {
  // this function moves the internal state of the item (checking against `now`, increasing counters, etc).
  match item.poll(&now) {
    NotReadyToRefetch => {},
    ReadyToRefetch => {
       asking.insert(*hash);
        if asking.len() > MAX_ASKING {
           break;
        }
    },
    // I've made up this case, but I really think we should actually remove items if there is no way
    // we can retrieve them. The previous code was pretty much never removing anything from `unfetched`
    // (except for the case when we actually retrieve the hash)
    RefetchTimeout => {
      remove_from_unfetched.push(*hash);
    }
}

for hash in remove_from_unfeched {
  self.unfetched_pooled_transactions.remove(&hash);
}
// avoid copying, we can just pass an iterator instead of a slice.
SyncRequester::request_pooled_transactions(self, io, peer_id, asking.iter());
```

Reply:
I have rewritten this part with additional comments, hopefully it's clear now. I could adopt your version if you still think it's bad though.